vmdb/app/models/mixins/process_tasks_mixin.rb

```diff
@@ -0,0 +1,123 @@
+      else
+        exception_if_task_unknown(options)
+        options[:userid] ||= "system"
+        invoke_tasks_queue(options)
+      end
+    end
+
+    def invoke_tasks_queue(options)
+      MiqQueue.put(:class_name => name, :method_name => "invoke_tasks", :args => [options])
+    end
+
+    # Performs tasks received from the UI via the queue
+    def invoke_tasks(options)
+      local, remote = partition_ids_by_remote_region(options[:ids])
+      invoke_tasks_local(options.merge(:ids => local)) unless local.empty?
+
+      # TODO: invoke_tasks_remote currently is only implemented by VmOrTemplate.
+      # it can be refactored to be generalized like invoke_tasks_local
+      invoke_tasks_remote(options.merge(:ids => remote)) if !remote.empty? && respond_to?("invoke_tasks_remote")
+    end
+
+    def invoke_tasks_local(options)
+      options[:invoke_by] = task_invoked_by(options)
+      args = task_arguments(options)
+
+      instances, tasks = validate_tasks(options)
+
+      instances.each_with_index do |instance, idx|
+        task = MiqTask.where(:id => tasks[idx]).first
+
+        if task && task.status == "Error"
+          task_audit_event(:failure, options, :target_id => instance.id, :message => task.message)
+          task.state_finished
+          next
+        end
+
+        invoke_task_local(task, instance, options, args)
+
+        msg = "#{instance.name}: '#{options[:task]}' initiated"
+        task_audit_event(:success, options, :target_id => instance.id, :message => msg)
+        task.update_status("Queued", "Ok", "Task has been queued") if task
+      end
+    end
+
+    # default: invoked by task, can be overridden
+    def task_invoked_by(_options)
+      :task
+    end
+
+    # default: only handles retirement, can be overridden
+    def task_arguments(options)
+      options[:task] == 'retire_now' ? [options[:userid]] : []
+    end
+
+    # default implementation, can be overridden
+    def invoke_task_local(task, instance, options, args)
+      cb = {
+        :class_name  => task.class.to_s,
+        :instance_id => task.id,
+        :method_name => :queue_callback,
+        :args        => ["Finished"]
+      } if task
+
+      MiqQueue.put(
+        :class_name   => name,
+        :instance_id  => instance.id,
+        :method_name  => options[:task],
+        :args         => args,
+        :miq_callback => cb
+      )
+    end
+
+    # Helper method for invoke_tasks, to determine the instances and the tasks associated
+    def validate_tasks(options)
+      tasks = []
+
+      instances = base_class.where(:id => options[:ids]).order("lower(name)").to_a
+      return instances, tasks unless options[:invoke_by] == :task # jobs will be used instead of tasks for feedback
+
+      instances.each do |instance|
+        # create a task for each instance
+        task = MiqTask.create(:name => "#{instance.name}: '#{options[:task]}'", :userid => options[:userid])
```

Comment:
If we are creating the task, why don't we just return that instead of the id...the caller ends up having to requery it, which isn't necessary technically.